rclpy/rclpy/executors.py

```diff
@@ -384,23 +399,41 @@ def wait_for_ready_callbacks(self, timeout_sec=None, nodes=None):
                         yield handler, srv, node
 
             # Check timeout timer
-            if (timeout_nsec == 0 or
-                    (timeout_timer is not None and timeout_timer.timer_pointer in timers_ready)):
-                break
+            if (
+                timeout_nsec == 0 or
+                (timeout_timer is not None and timeout_timer.timer_pointer in timers_ready)
+            ):
+                raise TimeoutException()
+
+    def wait_for_ready_callbacks(self, *args, **kwargs):
+        """
+        Reuse generator and return callbacks that are ready to be performed.
+
+        See :func:`Executor._wait_for_ready_callbacks` for documentation
+        """
+        if self._cb_iter is None or self._last_args != args or self._last_kwargs != kwargs:
+            # Create a new generator
+            self._last_args = args
+            self._last_kwargs = kwargs
+            self._cb_iter = self._wait_for_ready_callbacks(*args, **kwargs)
+        try:
+            return next(self._cb_iter)
+        except StopIteration:
+            # Ran out of work, recursively create a new generator
+            self._cb_iter = None
+            return self.wait_for_ready_callbacks(*args, **kwargs)
 
 
 class SingleThreadedExecutor(Executor):
     """Runs callbacks in the thread which calls :func:`SingleThreadedExecutor.spin`."""
 
-    def __init__(self):
-        super().__init__()
```

Comment:
Do we really want to remove this call to `Executor.__init__`?

Reply:
My understanding is that a method that is not overloaded in the child class means that the parent version of the method is used. I believe this applies to all methods included the `__init_` one

Reply:
I wasn't sure about this, so I just tried it out with the following program:

```
class Base:
	def __init__(self):
		print("Base init")

class Sub:
	def do_something(self):
		print("Sub do_something")

x = Sub()
x.do_something()
```

If you run that under python3, the output you get is just "Sub do_something"; you never see the "Base init" printed out.

Reply:
can you try the same thing with Sub inheriting from base ?
```
class Base:
	def __init__(self):
		print("Base init")

class Sub(Base):
	def do_something(self):
		print("Sub do_something")

x = Sub()
x.do_something()
```

Reply:
D'oh, sorry.  I forgot that bit.  Never mind, you are totally correct.